src/ipa_pc/mod.rs

```diff
@@ -110,21 +111,53 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
         for (labeled_commitment, value) in labeled_commitments.zip(values) {
             let commitment = labeled_commitment.commitment();
             combined_v += &(cur_challenge * &value);
-            combined_commitment_proj += &labeled_commitment.commitment().comm.mul(cur_challenge);
-            cur_challenge = opening_challenges(opening_challenge_counter);
-            opening_challenge_counter += 1;
 
-            let degree_bound = labeled_commitment.degree_bound();
-            assert_eq!(degree_bound.is_some(), commitment.shifted_comm.is_some());
+            let segments_count = commitment.comm.len();
+
+            if segments_count > 1 {
 
-            if let Some(degree_bound) = degree_bound {
-                let shift = point.pow([(vk.supported_degree() - degree_bound) as u64]);
-                combined_v += &(cur_challenge * &value * &shift);
-                combined_commitment_proj += &commitment.shifted_comm.unwrap().mul(cur_challenge);
+                let mut comm_lc = <G::Projective as ProjectiveCurve>::zero();
+                for (i, comm_single) in commitment.comm.iter().enumerate() {
+                    let is = i * key_len;
+                    comm_lc += &comm_single.mul(point.pow(&[is as u64]));
+                }
+                combined_commitment_proj += &comm_lc.mul(&cur_challenge);
+
+            } else {
+
+                combined_commitment_proj += &commitment.comm[0].mul(cur_challenge);
             }
 
             cur_challenge = opening_challenges(opening_challenge_counter);
             opening_challenge_counter += 1;
+
+            let degree_bound = labeled_commitment.degree_bound();
+
+            let degree_bound_len = degree_bound.and_then(|degree_bound_len| {
+                if (degree_bound_len + 1) % key_len != 0 { Some(degree_bound_len + 1) } else { None }
```

Comment:
As in the prover: 
we need to check that the degree_bound of a polynomial is "situated" in the last segment:
```
(segment_count-1) * segment_len < degree_bound <= segment_count * segment_len
```
Otherwise the commitment is ill-posed.

Reply:
This assertion added to the "check_degrees_and_bounds" function

Reply:
Fine, but we need to call this function in the verifier (!) at some place, otherwise a prover can trick the verifier on the degree bound. I'd propose to do that right here in the succinct_check(), and the same for the batched variant.

Reply:
Done